### ISSUE TITLE
Add alignment controls to grid layout

### ISSUE THREAD

**Issue by tellthemachines** [WordPress/gutenberg#62895]
Flex layouts such as used in Row and Stack blocks come with justification and alignment options for their content:

<img width="1015" alt="Screenshot 2024-06-27 at 11 59 21 AM" src="https://github.com/WordPress/gutenberg/assets/8096000/13df99ae-06f5-4354-bfc5-65ff951a2faf">

It would be good if Grid had something similar. The implementation should be very much like flex, the main thing I guess is to determine whether they should be toolbar or sidebar controls (flex is a bit awkward right now in that it has a mix of both).

Cc @luminuu who asked about this in a hallway hangout. Feel free to add to the issue if I left anything out!

**Comment by luminuu** [comment]
Thanks for opening the issue! I think this has to be separted into two parts: aligning the contents of the grid with `justify-items` and `align-items`, as well as aligning the rows and columns of the grid with `justify-content` and `align-content`.

To go even further, I think it should also be discussed if a child element can be positioned separately with `justify-self` and `align-self`.

**Comment by tellthemachines** [comment]
> as well as aligning the rows and columns of the grid with `justify-content` and `align-content`

I don't think `justify-content` will work with our current grid setup, because in both manual and auto the grid tracks occupy the full width of their container. `align-content` could be used in conjunction with min-height on a Grid block, but I'm not sure there's a use case for it that wouldn't also be covered by separating gap controls to be able to set vertical and horizontal gap to different values, or using padding. Did you have any specific designs in mind?

>  I think it should also be discussed if a child element can be positioned separately with justify-self and align-self.

This I think would be useful and worth looking into! It could be added to the child controls.

**Comment by tellthemachines** [comment]
@WordPress/gutenberg-design would you folks be able to help with design for these controls? We're looking at:

* For blocks with Grid layout: justification and alignment controls, similar to the Flex layout ones (can we just reuse those? Will the layout panel look too busy with these in addition to the current Grid controls? It might be time to look into #44560)

* For child blocks of a Grid layout: self-justification and self-alignment controls. I think it would make sense for these to be close to the current child layout controls (Column and Row span for Grids) but semantically it's weird to have justification and alignment under Dimensions. This means it's probably time to give consideration to something like #47902.

**Comment by richtabor** [comment]
I don’t think we should not wait on https://github.com/WordPress/gutenberg/issues/44560 or https://github.com/WordPress/gutenberg/issues/47902. 

Could the matrix potentially work here, in the toolbar? As a way to change justification and align together.

**Comment by jasmussen** [comment]
The missing piece in the matrix is the _space-between_ properties, though perhaps we can still start with just the basic matrix _without_ space-between, because space-between is both complex (therefore secondary and possibly a thing for the inspector), and could potentially be built into the block spacing control itself, to benefit both grid and flex based blocks.

**Comment by jasmussen** [comment]
@tellthemachines Looking at this one again, is there a design or a decision I can help you with on this one? Would a review of #79472 help?

**Comment by tellthemachines** [comment]
@jasmussen thanks for asking! I started working on it in https://github.com/WordPress/gutenberg/pull/79472 but ran into problems with the resize handles, because once we add alignments to grid items, the handles follow the item size and the whole resize computation (which depends on grid cells having fixed sizes) gets thrown off 😭 

So I parked it for a bit because I had higher priority stuff to look at (mostly responsive styles work). I'll pick it up again at some point, but not quite sure how best to approach it tbh. One thing that occurred to me was we could disable resizing when using alignments other than the default `stretch`. Happy to hear your ideas on this!

**Comment by jasmussen** [comment]
I'd agree it isn't very high priority, so I'm mostly looking at older issues in a triage sense to see if there's any low hanging fruit to pick.

One of the issues is, we already have resize handles, and for _some_ operations it works great, for others it shifts things around in a way that I find to be as confusing as the grid spec itself is:

<img width="1517" height="855" alt="Image" src="https://github.com/user-attachments/assets/62fad237-5768-4e58-a53d-da295c41fa27" />

I don't completely know how to solve that, other than possibly locking down resize controls more than they are today, really limiting how much you _can_ resize. So yes, I'd encourage disabling as much as possible under as many circumstances as possible, _in the canvas_. The fact that we do have in-canvas tools to help you I see as a great starting point, but I also suspect that the true grid customisation pieces will likely require more careful tinkering of elements in the inspector instead. 

A loose thought: @talldan made a really cool Table V2 concept recently where the innerblocks were really not surfaced as innerblocks, but rather as "cells". I don't know if that translates directly to this, but there was something quite nice about the simpler interface there that I think would go well with the Grid block too. Not sure there's an action item, sharing mostly in case it inspires ideas.

**Comment by tellthemachines** [comment]
Yeah I agree the resize handles in the default "auto" mode don't always provide a great experience; we should explore them further in "manual" mode for which they are much better suited. I wrote up a new tracking issue for it: #81796 and hope to be able to pick it up soon. For this one, maybe I'll go ahead and hide them if alignments are on.

**Comment by jasmussen** [comment]
Nice. Is this issue then still relevant or should it be closed in favour of 81796?
